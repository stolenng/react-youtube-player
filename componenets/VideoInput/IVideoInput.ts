export default interface IVideoInput{
    onSubmit: (id: string) => void;
}