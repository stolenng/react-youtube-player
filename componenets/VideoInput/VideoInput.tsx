import * as React from "react";

import IVideoInput from './IVideoInput';

type State = {
    value: string;
    class: string;
};

export default class VideoInputComponent extends React.Component<IVideoInput, State> {
    state: State = {
        value: '',
        class: ''
    };

    constructor(props: IVideoInput) {
        super(props);
    }

    handleChange = (event: React.FormEvent<EventTarget>) => {
        const target = event.target as HTMLInputElement;

        this.setState({value: target.value, class: '' });
    }
    
    handleSubmit = (event: React.FormEvent<EventTarget>) => {
        event.preventDefault();
        
        if (this.validate(this.state.value) && this.extractId(this.state.value)) {
            const id = this.extractId(this.state.value);
           
            this.props.onSubmit(id);
            this.reset();    
        }
        else {
            this.setState({value: this.state.value, class: 'error' });            
        }
    }

    reset() {
        this.setState({ value: '', class: '' });
    }
    
    validate(url: string): boolean {
        if (!url) {
            return false;
        }

        const regex = /^(http(s)?:\/\/)?((w){3}.)?youtu(be|.be)?(\.com)?\/.+/gm;     

        return regex.test(url);
    }

    getClass() : string {
        if (this.state.value) {
            return '';
        }

        return this.validate(this.state.value) ? '' : 'error'; 
    }

    extractId(url: string): string {
        const regex = /^.*(youtu.be\/|v\/|e\/|u\/\w+\/|embed\/|v=)([^#\&\?]*).*/;
        const result = url.match(regex);

        if (!result) {
            return null;
        }

        return result[result.length-1];
    }
     
    render() {
        return (
            <div className="video-input">
                <form onSubmit={this.handleSubmit}>
                    <input className={`input ${this.state.class}`} type="text" onChange={this.handleChange} value={this.state.value} placeholder="Insert valid youtube url"/>
                    <button className="button" type="submit">Add</button>
                </form>
            </div>
        );
    }
}