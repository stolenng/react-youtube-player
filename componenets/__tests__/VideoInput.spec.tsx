import * as React from "react";
import { shallow } from "enzyme";
import * as renderer from 'react-test-renderer';

import VideoInput from '../VideoInput/VideoInput';




describe('VideoInput', () => {
    it("Init input Componenet", () => {
        const tree = renderer
            .create(<VideoInput onSubmit={() => { }}></VideoInput>)
            .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('Test input validation', () => {
        const mockFn = jest.fn();
        
        const VideoInputComponent = shallow(
            <VideoInput onSubmit={mockFn}></VideoInput>
        );

        const url = 'https://www.youtube.com/watch?v=M3w2xDPHr90&list=RDkTRhXVq_8Nw&index=4';

        const instance = VideoInputComponent.instance() as VideoInput;

        expect(instance.validate(url)).toEqual(true);        
    });

    it('Test input id extraction', () => {
        const mockFn = jest.fn();
        
        const VideoInputComponent = shallow(
            <VideoInput onSubmit={mockFn}></VideoInput>
        );

        const url = 'https://www.youtube.com/watch?v=M3w2xDPHr90&list=RDkTRhXVq_8Nw&index=4';

        const instance = VideoInputComponent.instance() as VideoInput;

        expect(instance.extractId(url)).toEqual('M3w2xDPHr90');        
    });
});