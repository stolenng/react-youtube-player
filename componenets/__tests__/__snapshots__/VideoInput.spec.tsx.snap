// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoInput Init input Componenet 1`] = `
<div
  className="video-input"
>
  <form
    onSubmit={[Function]}
  >
    <input
      className="input "
      onChange={[Function]}
      placeholder="Insert valid youtube url"
      type="text"
      value=""
    />
    <button
      className="button"
      type="submit"
    >
      Add
    </button>
  </form>
</div>
`;
